File: app/src/main/java/de/hdc/kspchecklist/ItemDetailAdapter.java
package de.hdc.kspchecklist;

import android.content.Context;
import android.graphics.Color;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;

import de.hdc.kspchecklist.data.CheckListItem;
import de.hdc.kspchecklist.data.DataIO;

/**
 * Created by DerTroglodyt on 2017-02-18 09:33.
 * Email dertroglodyt@gmail.com
 * Copyright by HDC, Germany
 */

class ItemDetailAdapter extends RecyclerView.Adapter<ItemDetailAdapter.ViewHolder>
        implements View.OnClickListener {   //}, View.OnCreateContextMenuListener {

    ItemDetailAdapter(Context context, String fileName, ArrayList<CheckListItem> objects) {
//        super(context, 0, objects);
        this.context = context;
        this.fileName = fileName;
        this.list = objects;
    }

    // Create new views (invoked by the layout manager)
    @Override
    public ItemDetailAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        // create a new view
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_detail_content, parent, false);
        //CheckBox v = (CheckBox) test.findViewById(R.id.checkbox);
//        CheckBox v = (CheckBox) LayoutInflater.from(parent.getContext())
//                .inflate(R.layout.item_detail_content, parent, false);
        // set the view's size, margins, paddings and layout parameters
        return new ViewHolder(v, this);
    }

    // Replace the contents of a view (invoked by the layout manager)
    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        // - get element from your dataset at this position
        // - replace the contents of the view with that element
        holder.cb.setText(list.get(position).name);
        holder.cb.setChecked(list.get(position).checked);
        holder.cb.setTag(position);
        setBGColor(holder.cb, false);
    }

    // Return the size of your dataset (invoked by the layout manager)
    @Override
    public int getItemCount() {
        return list.size();
    }

    void remove(int position) {
        list.remove(position);
        try {
            DataIO.writeLocalFile(context, fileName, list);
        } catch (IOException e) {
            e.printStackTrace();
        }
        notifyItemRemoved(position);
    }

    void swap(int firstPosition, int secondPosition){
        Collections.swap(list, firstPosition, secondPosition);
        try {
            DataIO.writeLocalFile(context, fileName, list);
        } catch (IOException e) {
            e.printStackTrace();
        }
        notifyItemMoved(firstPosition, secondPosition);
    }

    void setBGColor(CheckBox cb, boolean isDragging) {
        if (isDragging) {
            cb.setBackgroundColor(Color.CYAN);
        } else {
            if (cb.isChecked()) {
                cb.setBackgroundColor(Color.argb(64, 0, 255, 0));
            } else {
                cb.setBackgroundColor(Color.argb(128, 255, 255, 0));
            }
        }
    }

    @Override
    public void onClick(View view) {
        CheckBox cb = (CheckBox) view;
        setBGColor(cb, false);
        CheckListItem cli = CheckListItem.create(cb.getText().toString(), cb.isChecked());
        Integer i = (Integer) view.getTag();
        list.set(i, cli);
        try {
            DataIO.writeLocalFile(context, fileName, list);
        } catch (IOException e) {
            e.printStackTrace();
        }
//        notifyDataSetChanged();
    }

//    @Override
//    public void onCreateContextMenu(final ContextMenu menu, final View v, final ContextMenu.ContextMenuInfo menuInfo) {
//        if (menuInfo == null ) {
//            return;
//        }
//        AdapterView.AdapterContextMenuInfo info = (AdapterView.AdapterContextMenuInfo) menuInfo;
//        menu.setHeaderTitle(list.get(info.position).name);
//        String[] menuItems = context.getResources().getStringArray(R.array.menu_detail);
//        for (int i = 0; i<menuItems.length; i++) {
//            menu.add(Menu.NONE, i, i, menuItems[i]);
//        }
//    }

    // Provide a reference to the views for each data item
    // Complex data items may need more than one view per item, and
    // you provide access to all the views for a data item in a view holder
    static class ViewHolder extends RecyclerView.ViewHolder {
        // each data item is just a string in this case
        final CheckBox cb;

        ViewHolder(View v, ItemDetailAdapter adapter) {
            super(v);
            cb = (CheckBox) v.findViewById(R.id.checkbox);
            cb.setOnClickListener(adapter);
//            cb.setOnCreateContextMenuListener(adapter);
        }
    }

    private final Context context;
    private final String fileName;
    private final ArrayList<CheckListItem> list;
}

File: app/src/main/java/de/hdc/kspchecklist/ItemDetailAdapter.kt
package de.hdc.kspchecklist

import android.content.Context
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import androidx.recyclerview.widget.RecyclerView
import de.hdc.kspchecklist.data.CheckListPersistenceSource
import de.hdc.kspchecklist.domain.CheckListItem
import java.io.IOException
import java.util.*

/**
 * Created by DerTroglodyt on 2017-02-18 09:33.
 * Email dertroglodyt@gmail.com
 * Copyright by HDC, Germany
 */

internal class ItemDetailAdapter(
    private val context: Context,
    private val persistence: CheckListPersistenceSource,
    private val fileName: String,
    private val list: ArrayList<CheckListItem>
) : RecyclerView.Adapter<ItemDetailAdapter.ViewHolder>(), View.OnClickListener {

    // Create new views (invoked by the layout manager)
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemDetailAdapter.ViewHolder {
        // create a new view
        val v = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_detail_content, parent, false)
        //CheckBox v = (CheckBox) test.findViewById(R.id.checkbox);
        //        CheckBox v = (CheckBox) LayoutInflater.from(parent.getContext())
        //                .inflate(R.layout.item_detail_content, parent, false);
        // set the view's size, margins, paddings and layout parameters
        return ViewHolder(v, this)
    }

    // Replace the contents of a view (invoked by the layout manager)
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        // - get element from your dataset at this position
        // - replace the contents of the view with that element
        holder.cb.text = list[position].name
        holder.cb.isChecked = list[position].checked
        holder.cb.tag = position
        setBGColor(holder.cb, false)
    }

    // Return the size of your dataset (invoked by the layout manager)
    override fun getItemCount(): Int {
        return list.size
    }

    fun remove(position: Int) {
        list.removeAt(position)
        try {
            persistence.saveList(fileName, list)
        } catch (e: IOException) {
            e.printStackTrace()
        }

        notifyItemRemoved(position)
    }

    fun swap(firstPosition: Int, secondPosition: Int) {
        Collections.swap(list, firstPosition, secondPosition)
        try {
            persistence.saveList(fileName, list)
        } catch (e: IOException) {
            e.printStackTrace()
        }

        notifyItemMoved(firstPosition, secondPosition)
    }

    fun setBGColor(cb: CheckBox, isDragging: Boolean) {
        if (isDragging) {
            cb.setBackgroundColor(Color.CYAN)
        } else {
            if (cb.isChecked) {
                cb.setBackgroundColor(Color.argb(64, 0, 255, 0))
            } else {
                cb.setBackgroundColor(Color.argb(128, 255, 255, 0))
            }
        }
    }

    override fun onClick(view: View) {
        val cb = view as CheckBox
        setBGColor(cb, false)
        val cli = CheckListItem(cb.text.toString(), cb.isChecked)
        val i = view.getTag() as Int
        list[i] = cli
        try {
            persistence.saveList(fileName, list)
        } catch (e: IOException) {
            e.printStackTrace()
        }

        //        notifyDataSetChanged();
    }

    //    @Override
    //    public void onCreateContextMenu(final ContextMenu menu, final View v, final ContextMenu.ContextMenuInfo menuInfo) {
    //        if (menuInfo == null ) {
    //            return;
    //        }
    //        AdapterView.AdapterContextMenuInfo info = (AdapterView.AdapterContextMenuInfo) menuInfo;
    //        menu.setHeaderTitle(list.get(info.position).name);
    //        String[] menuItems = context.getResources().getStringArray(R.array.menu_detail);
    //        for (int i = 0; i<menuItems.length; i++) {
    //            menu.add(Menu.NONE, i, i, menuItems[i]);
    //        }
    //    }

    // Provide a reference to the views for each data item
    // Complex data items may need more than one view per item, and
    // you provide access to all the views for a data item in a view holder
    internal class ViewHolder(v: View, adapter: ItemDetailAdapter) : RecyclerView.ViewHolder(v) {
        // each data item is just a string in this case
        val cb: CheckBox = v.findViewById<View>(R.id.checkbox) as CheckBox

        init {
            cb.setOnClickListener(adapter)
            //            cb.setOnCreateContextMenuListener(adapter);
        }
    }
}
